File: go/test/todo_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/korean-json-sdk"
	"github.com/voxgig-sdk/korean-json-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestTodoEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Todo(nil)
		if ent == nil {
			t.Fatal("expected non-nil TodoEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := todoBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "todo." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set KOREANJSON_TEST_TODO_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		todoRef01Ent := client.Todo(nil)
		todoRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "todo"}, setup.data), "todo_ref01"))

		todoRef01DataResult, err := todoRef01Ent.Create(todoRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		todoRef01Data = core.ToMapAny(todoRef01DataResult)
		if todoRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if todoRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		todoRef01Match := map[string]any{}

		todoRef01ListResult, err := todoRef01Ent.List(todoRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		todoRef01List, todoRef01ListOk := todoRef01ListResult.([]any)
		if !todoRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", todoRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(todoRef01List), map[string]any{"id": todoRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		todoRef01DataUp0Up := map[string]any{
			"id": todoRef01Data["id"],
		}

		todoRef01MarkdefUp0Name := "title"
		todoRef01MarkdefUp0Value := fmt.Sprintf("Mark01-todo_ref01_%d", setup.now)
		todoRef01DataUp0Up[todoRef01MarkdefUp0Name] = todoRef01MarkdefUp0Value

		todoRef01ResdataUp0Result, err := todoRef01Ent.Update(todoRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		todoRef01ResdataUp0 := core.ToMapAny(todoRef01ResdataUp0Result)
		if todoRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if todoRef01ResdataUp0["id"] != todoRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if todoRef01ResdataUp0[todoRef01MarkdefUp0Name] != todoRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", todoRef01MarkdefUp0Name, todoRef01ResdataUp0[todoRef01MarkdefUp0Name])
		}

		// LOAD
		todoRef01MatchDt0 := map[string]any{
			"id": todoRef01Data["id"],
		}
		todoRef01DataDt0Loaded, err := todoRef01Ent.Load(todoRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		todoRef01DataDt0LoadResult := core.ToMapAny(todoRef01DataDt0Loaded)
		if todoRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if todoRef01DataDt0LoadResult["id"] != todoRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		todoRef01MatchRm0 := map[string]any{
			"id": todoRef01Data["id"],
		}
		_, err = todoRef01Ent.Remove(todoRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		todoRef01MatchRt0 := map[string]any{}

		todoRef01ListRt0Result, err := todoRef01Ent.List(todoRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		todoRef01ListRt0, todoRef01ListRt0Ok := todoRef01ListRt0Result.([]any)
		if !todoRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", todoRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(todoRef01ListRt0), map[string]any{"id": todoRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func todoBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "todo", "TodoTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read todo test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse todo test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"todo01", "todo02", "todo03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("KOREANJSON_TEST_TODO_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"KOREANJSON_TEST_TODO_ENTID": idmap,
		"KOREANJSON_TEST_LIVE":      "FALSE",
		"KOREANJSON_TEST_EXPLAIN":   "FALSE",
		"KOREANJSON_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["KOREANJSON_TEST_TODO_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["KOREANJSON_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["KOREANJSON_APIKEY"],
			},
			extra,
		})
		client = sdk.NewKoreanJsonSDK(core.ToMapAny(mergedOpts))
	}

	live := env["KOREANJSON_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["KOREANJSON_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/todo_entity_test.rb
# Todo entity test

require "minitest/autorun"
require "json"
require_relative "../KoreanJson_sdk"
require_relative "runner"

class TodoEntityTest < Minitest::Test
  def test_create_instance
    testsdk = KoreanJsonSDK.test(nil, nil)
    ent = testsdk.Todo(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = todo_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "todo." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set KOREANJSON_TEST_TODO_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    todo_ref01_ent = client.Todo(nil)
    todo_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.todo"), "todo_ref01"))

    todo_ref01_data_result, err = todo_ref01_ent.create(todo_ref01_data, nil)
    assert_nil err
    todo_ref01_data = Helpers.to_map(todo_ref01_data_result)
    assert !todo_ref01_data.nil?
    assert !todo_ref01_data["id"].nil?

    # LIST
    todo_ref01_match = {}

    todo_ref01_list_result, err = todo_ref01_ent.list(todo_ref01_match, nil)
    assert_nil err
    assert todo_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(todo_ref01_list_result),
      { "id" => todo_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    todo_ref01_data_up0_up = {
      "id" => todo_ref01_data["id"],
    }

    todo_ref01_markdef_up0_name = "title"
    todo_ref01_markdef_up0_value = "Mark01-todo_ref01_#{setup[:now]}"
    todo_ref01_data_up0_up[todo_ref01_markdef_up0_name] = todo_ref01_markdef_up0_value

    todo_ref01_resdata_up0_result, err = todo_ref01_ent.update(todo_ref01_data_up0_up, nil)
    assert_nil err
    todo_ref01_resdata_up0 = Helpers.to_map(todo_ref01_resdata_up0_result)
    assert !todo_ref01_resdata_up0.nil?
    assert_equal todo_ref01_resdata_up0["id"], todo_ref01_data_up0_up["id"]
    assert_equal todo_ref01_resdata_up0[todo_ref01_markdef_up0_name], todo_ref01_markdef_up0_value

    # LOAD
    todo_ref01_match_dt0 = {
      "id" => todo_ref01_data["id"],
    }
    todo_ref01_data_dt0_loaded, err = todo_ref01_ent.load(todo_ref01_match_dt0, nil)
    assert_nil err
    todo_ref01_data_dt0_load_result = Helpers.to_map(todo_ref01_data_dt0_loaded)
    assert !todo_ref01_data_dt0_load_result.nil?
    assert_equal todo_ref01_data_dt0_load_result["id"], todo_ref01_data["id"]

    # REMOVE
    todo_ref01_match_rm0 = {
      "id" => todo_ref01_data["id"],
    }
    _, err = todo_ref01_ent.remove(todo_ref01_match_rm0, nil)
    assert_nil err

    # LIST
    todo_ref01_match_rt0 = {}

    todo_ref01_list_rt0_result, err = todo_ref01_ent.list(todo_ref01_match_rt0, nil)
    assert_nil err
    assert todo_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(todo_ref01_list_rt0_result),
      { "id" => todo_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def todo_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "todo", "TodoTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = KoreanJsonSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["todo01", "todo02", "todo03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["KOREANJSON_TEST_TODO_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "KOREANJSON_TEST_TODO_ENTID" => idmap,
    "KOREANJSON_TEST_LIVE" => "FALSE",
    "KOREANJSON_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["KOREANJSON_TEST_TODO_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["KOREANJSON_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = KoreanJsonSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["KOREANJSON_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["KOREANJSON_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
